src/main/ruby/truffleruby/core/truffle/polyglot.rb

```diff
@@ -34,6 +34,157 @@ def self.as_enumerable(object)
     Truffle::Interop.enumerable(object)
   end
 
+  # HasArrayElementsTrait
+  module HasArrayElementsTrait
+
+    private def read_array_element(index)
+      Truffle::Interop.read_array_element(self, index)
+    end
+
+    def [](index)
+      index += length if index < 0
+      return nil if (index < 0 || index >= length)
+      read_array_element(index)
+    end
+
+    alias_method :at, :[]
+
+    def first
+      read_array_element(0)
+    end
+
+    def last
+      read_array_element(length - 1)
+    end
+
+    def size
+      Truffle::Interop.array_size(self)
+    end
+
+    alias_method :length, :size
+
+    def count(item = nil)
+      seq = 0
+      if (!item.nil?)
+        each { |o| seq += 1 if item == o }
+      elsif block_given?
+        each { |o| seq += 1 if yield(o) }
+      else
+        return size
+      end
+      seq
+    end
+
+    def to_a
+      if Truffle::Interop.has_array_elements?(self)
```

Comment:
I wouldn't check here, if this trait is used, then the foreign should already have hasArrayElements()==true.